### ISSUE TITLE
Enable forward to return multiple tensors

### ISSUE THREAD

**Issue by SMHendryx** [LaurentMazare/tch-rs#298]
I could be missing it but as far as I can tell the `forward_ts` method on `CModule` only supports returning a single tensor. This is not always possible or ideal such as when your model predicts multiple tensors of different shapes so that we can't concatenate or stack them into a single tensor. (In theory we could flatten everything to a vector, concat along the single axis and then reshape in rust, but this is a huge hassle and if the return tensors are dynamically sized this becomes even more inconvenient.) Ideally, forward should be able to return a tuple or vector of tensors. This is possible in the torch C++ API: https://github.com/pytorch/pytorch/issues/15523#issuecomment-456688590

If we try to return a tuple of tensors this ends up throwing this runtime error: https://github.com/LaurentMazare/tch-rs/blob/04203e929925a92ad99d91d33ad42256398c9d06/torch-sys/libtch/torch_api.cpp#L828

This is reproducible with:
`convert_model_to_torchscript.py`:
```
from typing import Optional, Tuple

import torch
from torch import nn


class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.conv_branch_0 = nn.Conv2d(3, 64, kernel_size=3, padding=1)
        self.conv_branch_1 = nn.Conv2d(3, 64, kernel_size=3, padding=1)

    def forward(self, inputs) -> Tuple[torch.Tensor, torch.Tensor]:
        x0: torch.Tensor = self.conv_branch_0(inputs)
        x1: torch.Tensor = self.conv_branch_1(inputs)
        x1 = x1.mean(dim=(-1, -2))
        return x0, x1


def convert(model: torch.nn.Module, example: Optional[torch.Tensor] = None, input_dims: Tuple[int] = (224, 224)):
    # An example input you would normally provide to your model's forward() method.
    if example is None:
        example = torch.rand(1, 3, *input_dims)

    # Use torch.jit.trace to generate a torch.jit.ScriptModule via tracing.
    traced_script_module = torch.jit.trace(model, example)

    return traced_script_module


def main():
    model = Model()
    model.eval()
    model = convert(model)
    model.save("forward_multi_tens_model.pt")


if __name__ == "__main__":
    main()
```

`main.rs`:
```
extern crate tch;
use anyhow::{bail, Result};
use tch::vision::imagenet;

pub fn main() -> Result<()> {
    let args: Vec<_> = std::env::args().collect();
    let (model_file, image_file) = match args.as_slice() {
        [_, m, i] => (m.to_owned(), i.to_owned()),
        _ => bail!("usage: main forward_multi_tens_model.pt image.jpg"),
    };
    // Load the image file and resize it
    let image = imagenet::load_image_and_resize224(image_file)?;

    // Load the Python saved module.
    let model = tch::CModule::load(model_file)?;

    let output = model.forward_ts(&[image.unsqueeze(0)])?;
    //    println!("Output tensor 0 shape: {:?}", output.0.size());
    //    println!("Output tensor 1 shape: {:?}", output.1.size());
    Ok(())
}
```

Thanks in advance!
-Sean

**Comment by LaurentMazare** [comment]
Hello,
The `forward_ts` method indeed returns a single tensor. However there is a `forward_is` method that operates on [ivalues](https://docs.rs/tch/0.3.0/tch/enum.IValue.html) and could be used on your example, e.g. via the following:
```rust
extern crate tch;
use anyhow::{bail, Result};
use tch::vision::imagenet;

pub fn main() -> Result<()> {
    let args: Vec<_> = std::env::args().collect();
    let (model_file, image_file) = match args.as_slice() {
        [_, m, i] => (m.to_owned(), i.to_owned()),
        _ => bail!("usage: main forward_multi_tens_model.pt image.jpg"),
    };
    // Load the image file and resize it
    let image = imagenet::load_image_and_resize224(image_file)?;

    // Load the Python saved module.
    let model = tch::CModule::load(model_file)?;

    let output = model.forward_is(&[tch::IValue::Tensor(image.unsqueeze(0))])?;
    println!("{:?}", output);
    Ok(())
}
```

**Comment by SMHendryx** [comment]
Thanks @LaurentMazare. That works. What is the appropriate way to index into the output `IValue` so that we can put the tensors into their own variables?

**Comment by LaurentMazare** [comment]
You can do some pattern matching to extract the different layers, for example:
```rust
    let (tensor1, tensor2) = match output {
        IValue::Tuple(ivalues) => match &ivalues[..] {
            [IValue::Tensor(t1), IValue::Tensor(t2)] => (t1.shallow_clone(), t2.shallow_clone()),
            _ => bail!("unexpected output {:?}", ivalues),
        },
        _ => bail!("unexpected output {:?}", output),
    };
    println!("Output tensor 0 shape: {:?}", tensor1.size());
    println!("Output tensor 1 shape: {:?}", tensor2.size());
```

Alternatively, you may want to create some helper functions, e.g. `extract_tuple(IValue) -> Result<Vec<IValue>>`, and `extract_tensor` and combine these.

**Comment by SMHendryx** [comment]
That works! Thanks for the quick reply @LaurentMazare. Closing this issue.
